---
title: "add_kgen50_Apr2017"
output: html_document
---

started: Alexey Larionov, 31Jan2017  
last updated: Alexey Larionov, 03Apr2017

# Summary

Adds wecare-nfe datawith kgen50 data (50 cases of different continental ancestory)  
Merged data will be used to generate ethnic-relevant set of eigenvectors  

The kgen data were generated in wecare-stat-Jan2017 analysis. 

Wecare data include output generated by the following wes pipeline versions:  
1) Alignment to b37 of Oct2015  
2) Variant calling of Nov2016  
3) Variant filtering and annotation of Jan2017
The wecare samples were selected as in the respective r01, r02, r03 r04 scripts in this folder.  

Input data: 275,516 vars x 678 cases (480 wecare + 198 nfe)  
Output data: 275,516 vars x 728 cases (480 wecare + 198 nfe + 50 kgen50)  

# start_section

```{r start_section}

# Time stamp
Sys.time()

# Folders
#/analysis/mtgroup_share
#/scratch/medgen
setwd("/analysis/mtgroup_share/scripts/wecare_stat_04.17/scripts")
source_data_folder <- "/analysis/mtgroup_share/scripts/wecare_stat_04.17/source_data"
interim_data_folder <- "/analysis/mtgroup_share/scripts/wecare_stat_04.17/interim_data"
results_folder <- "/analysis/mtgroup_share/scripts/wecare_stat_04.17/results"

```

# read_data

103,045 x 764  

103,045 variants (overlap of wecare-nfe and kgen50)  

Number of samples in raw data:  
760 = 512 wecare + 198 nfe + 50 kgen  
and 4 additional columns: Chrom, Pos, kgen-id, wecare-nfe-id  

```{r read_data}

# load wecare-nfe data
load(paste(interim_data_folder, "r04_filter_cases_and_variants_Apr2017.RData", sep="/"))

# Load kgen data

kgen50_cases.df <- read.table(paste(source_data_folder, "kgen_selected_50_samples.txt", sep="/"))
dim(kgen50_cases.df)

kgen50_vcf.df <- read.table(
  paste(source_data_folder, "wecare_nfe_kgen50_filt_vcf.txt", sep="/"), 
  header=TRUE, sep="\t", quote="")
dim(kgen50_vcf.df)

kgen50_gt.df <- read.table(
  paste(source_data_folder, "wecare_nfe_kgen50_filt_gt.txt", sep="/"), 
  header=TRUE, sep="\t", quote="")
dim(kgen50_gt.df)

kgen50_gq.df <- read.table(
  paste(source_data_folder, "wecare_nfe_kgen50_filt_gq.txt", sep="/"), 
  header=TRUE, sep="\t", quote="")
dim(kgen50_gq.df)

kgen50_dp.df <- read.table(
  paste(source_data_folder, "wecare_nfe_kgen50_filt_dp.txt", sep="/"), 
  header=TRUE, sep="\t", quote="")
dim(kgen50_dp.df)

```

# add_SplitVarIDs_to_rownames

SplitVarId is the variants id from wecare-nfe  

```{r add_SplitVarIDs_to_rownames}

# Set rownames to SplitVarId (the VarId from wecare-nfe)
rownames(kgen50_vcf.df) <- as.vector(kgen50_vcf.df$SplitVarID)
rownames(kgen50_gt.df) <- as.vector(kgen50_gt.df$SplitVarID)
rownames(kgen50_gq.df) <- as.vector(kgen50_gq.df$SplitVarID)
rownames(kgen50_dp.df) <- as.vector(kgen50_dp.df$SplitVarID)

# Check the consistency of rows
sum(rownames(kgen50_vcf.df) != rownames(kgen50_gt.df))
sum(rownames(kgen50_vcf.df) != rownames(kgen50_gq.df))
sum(rownames(kgen50_vcf.df) != rownames(kgen50_dp.df))

```

# check_kgen50_cases_consistency_select_cases_and_update_cases_names

Initial cases: 
760 = 512 wecare + 198 nfe + 50 kgen  

Selected cases:  
50 kgen  

```{r check_kgen50_cases_consistency_select_cases_and_update_cases_names}

# Update kgen50_cases table
colnames(kgen50_cases.df) <- c("case", "subpopulation", "population", "gender")
kgen50_cases.df$case <- paste(as.vector(kgen50_cases.df$case), "_kgen", sep="")
rownames(kgen50_cases.df) <- kgen50_cases.df$case
str(kgen50_cases.df)

# Check consistency of cases in gt, gq and dp
sum( sub(".GT$", "", colnames(kgen50_gt.df)) != sub(".GQ$", "", colnames(kgen50_gq.df)) )
sum( sub(".GT$", "", colnames(kgen50_gt.df)) != sub(".DP$", "", colnames(kgen50_dp.df)) )

# Edit cases names in gt, gq and dp tables
cases_raw <- colnames(kgen50_gt.df)
cases_renamed <- sub(".variant2.GT", "_kgen", cases_raw)
cases_renamed <- sub(".variant.GT", "", cases_renamed)

cases_renamed -> colnames(kgen50_gt.df)
cases_renamed -> colnames(kgen50_gq.df)
cases_renamed -> colnames(kgen50_dp.df)

# Check counts of cases
kgen50_cases <- grep(".variant2.GT", cases_raw, value=TRUE)
length(kgen50_cases) # 50

wecare_nfe_cases <- grep(".variant.GT", cases_raw, value=TRUE)
length(wecare_nfe_cases) # 710

wecare_cases <- grep("^P", wecare_nfe_cases, value=TRUE)
length(wecare_cases) # 512

nfe_cases <- grep("^[H,N]", wecare_nfe_cases, value=TRUE)
length(nfe_cases) # 198

# Check kgen50 and nfe overlap
kgen50_cases <- sub(".variant2.GT","",kgen50_cases)
nfe_cases <- sub(".variant.GT","",nfe_cases)

sum(kgen50_cases %in% nfe_cases) # 9
# One kgen50 EUR (HG01704) was removed from nfe set
# because of some odd QC metrics during our internal re-alignment

# Select cases
#selected_cases <- c(kgen50_cases, filtered_wecare_nfe_cases)
selected_cases <- kgen50_cases.df$case
length(selected_cases) # 728 = 50 + 198 + 480

# Update gt, gq and dp tables
kgen50_gt.df <- kgen50_gt.df[,selected_cases]

kgen50_gq.df <- kgen50_gq.df[,selected_cases]
kgen50_dp.df <- kgen50_dp.df[,selected_cases]

# No NA data in genotypes (as expected)
sum(is.na(kgen50_gt.df)) # 0

# No data in gq and dp for kgen50 only (as expected)
sum(!is.na(kgen50_gq.df)) # 0
sum(!is.na(kgen50_dp.df)) # 0

# Clean-up
rm(cases_raw, cases_renamed, kgen50_cases, wecare_nfe_cases, wecare_cases, nfe_cases, selected_cases, kgen50_gq.df, kgen50_dp.df)

```

# remove_multiallelic_variants_in_kgen50

103,045 -> 102,708 variants  

```{r remove_multiallelic_variants_in_kgen50}

ma <- grepl(",", as.vector(kgen50_vcf.df$ALT))

sum(ma)

kgen50_gt.df <- kgen50_gt.df[!ma,]
kgen50_vcf.df <- kgen50_vcf.df[!ma,]

dim(kgen50_gt.df)
dim(kgen50_vcf.df)

rm(ma)

```

# convert_kgen50_gt_data_frames_to_matrix

```{r convert_kgen50_gt_data_frames_to_matrix}

# data frames to matrices

class(kgen50_gt.df)

kgen50_gt.mx <- as.matrix(kgen50_gt.df)

class(kgen50_gt.mx)
str(kgen50_gt.mx)

# Clean-up
rm(kgen50_gt.df)

```

# recode_kgen50_gt_to_additive

All phased genotypes, no missed data

```{r recode_kgen50_gt_to_additive}

# --- Prepare genotypes vectors --- #

# Ref and alt
ref <- as.vector(kgen50_vcf.df$REF)
alt <- as.vector(kgen50_vcf.df$ALT)
ref[1:5]
alt[1:5]

# Homozygous reference
ref_ref_phased <- paste(ref, ref, sep="|")
ref_ref_phased[1:5]

# Heterozygous
alt_ref_phased <- paste(alt, ref, sep="|")
ref_alt_phased <- paste(ref, alt, sep="|")
alt_ref_phased[1:5]
ref_alt_phased[1:5]

# Homozygous alt (biallelic sites only!)
alt_alt_phased <- paste(alt, alt, sep="|")
alt_alt_phased[1:5]

# --- Convert genotypes to additive --- #

# Homozygous reference
0 -> kgen50_gt.mx[kgen50_gt.mx == ref_ref_phased]

# Heterozygous
1 -> kgen50_gt.mx[kgen50_gt.mx == alt_ref_phased]
1 -> kgen50_gt.mx[kgen50_gt.mx == ref_alt_phased]

# Homozygous alt (biallelic sites only!)
2 -> kgen50_gt.mx[kgen50_gt.mx == alt_alt_phased]

# Check the re-coding result
summary(as.factor(kgen50_gt.mx))

# Make sure the matrix is numeric
rownames_gt <- rownames(kgen50_gt.mx)
colnames_gt <- colnames(kgen50_gt.mx)
kgen50_gt.mx <- matrix(as.numeric(kgen50_gt.mx), nrow=nrow(kgen50_gt.mx))
rownames_gt -> rownames(kgen50_gt.mx)
colnames_gt -> colnames(kgen50_gt.mx)

min(kgen50_gt.mx, na.rm=TRUE)
max(kgen50_gt.mx, na.rm=TRUE)

# Check dimentions of gt
dim(kgen50_gt.mx)

# Clean-up
rm(ref, alt, ref_ref_phased, alt_ref_phased, ref_alt_phased, alt_alt_phased, rownames_gt, colnames_gt)

```

# remove_variants_with_the_uniform_genotypes_accross_all_kgen50_samples

102,708 -> 101,622 variants 

```{r remove_variants_with_the_uniform_genotypes_accross_all_kgen50_samples}

# Function to detect uniform numeric vector
uniform_vector.fnc <- function(x){
  if(min(x, na.rm=TRUE) == max(x, na.rm=TRUE)){return(TRUE)} else {return(FALSE)}}

# Variants with uniform genotypes accross all samples 
uniform_genotypes <- apply(kgen50_gt.mx, 1, uniform_vector.fnc)
summary(uniform_genotypes)
sum(uniform_genotypes)

# Remove variants with uniform genotypes accross all samples
kgen50_gt.mx <- kgen50_gt.mx[!uniform_genotypes,]
kgen50_vcf.df <- kgen50_vcf.df[!uniform_genotypes,]

# Check new size of data
dim(kgen50_gt.mx)
dim(kgen50_vcf.df)

# Clean-up
rm(uniform_vector.fnc, uniform_genotypes)

```

# estimate_numbers_of_variants_with_extreme_AFs

Apparent excess of group-A variants (AF>95%) over the group-B (AF<5%) reflects preferential selection of variants with rare allele in reference genome: why would it happen?  

```{r estimate_numbers_of_variants_with_extreme_AFs}

sum(as.vector(kgen50_vcf.df$AF) > 0.95) # rare variants with rare allele in reference genome (group A)
# 29,563

sum(as.vector(kgen50_vcf.df$AF) < 0.05) # rare variants with common allele in reference genome (group B)
# 19,462

```

# merge_kgen50_with_wicare-nfe

Cases:  
728 = 50 kgen  + 198 nfe + 480 wecare

```{r merge_kgen50_with_wicare-nfe}

# --- Select kgen50 variants overlapping with wecare-nfe --- #

# 101,622 -> 80,963

kgen50_split_var_ids <- as.vector(kgen50_vcf.df$SplitVarID)
wecare_nfe_split_var_ids <- as.vector(wecare_nfe_variants.df$SplitVarID)
overlap_split_var_ids <- intersect(kgen50_split_var_ids, wecare_nfe_split_var_ids)
length(overlap_split_var_ids) # 80,963

kgen50_gt.mx <- kgen50_gt.mx[overlap_split_var_ids,]
dim(kgen50_gt.mx)

kgen50_vcf.df <- kgen50_vcf.df[overlap_split_var_ids,]
dim(kgen50_vcf.df)

variants.df <- cbind(wecare_nfe_variants.df, rep(FALSE,nrow(wecare_nfe_variants.df)))
colnames(variants.df) <- c(colnames(wecare_nfe_variants.df), "kgen50")
str(variants.df)

TRUE -> variants.df[overlap_split_var_ids,"kgen50"]
sum(variants.df$kgen50) # 

# --- Merge genotypes --- #

kgen50_gt_2add.mx <- matrix(rep(NA,nrow(wecare_nfe_genotypes.mx)*50), nrow=nrow(wecare_nfe_genotypes.mx))
rownames(kgen50_gt_2add.mx) <- rownames(wecare_nfe_genotypes.mx)
colnames(kgen50_gt_2add.mx) <- colnames(kgen50_gt.mx)

for(var in rownames(kgen50_gt.mx)){kgen50_gt.mx[var,] -> kgen50_gt_2add.mx[var,]}

genotypes.mx <- cbind(kgen50_gt_2add.mx, wecare_nfe_genotypes.mx)
dim(genotypes.mx)

# --- Merge phenotypes --- #

phenotypes.df <- cbind(wecare_nfe_phenotypes.df, matrix(as.character(rep(NA,nrow(wecare_nfe_phenotypes.df)*3)), ncol=3))
colnames(phenotypes.df) <- c(colnames(wecare_nfe_phenotypes.df),c("subpopulation","population","gender"))
phenotypes.df$subpopulation <- as.character(phenotypes.df$subpopulation)
phenotypes.df$population <- as.character(phenotypes.df$population)
phenotypes.df$gender <- as.character(phenotypes.df$gender)
str(phenotypes.df)

kgen50_ph_2add.df <- phenotypes.df[1:50,]
NA -> kgen50_ph_2add.df[1:50,1:32]
str(kgen50_ph_2add.df)

rownames(kgen50_ph_2add.df) <- rownames(kgen50_cases.df)

for(cs in rownames(kgen50_cases.df)){
  kgen50_cases.df[cs, "case"] -> kgen50_ph_2add.df[cs, "wes_id"]
  -2 -> kgen50_ph_2add.df[cs, "cc"]
  as.character(kgen50_cases.df[cs, "subpopulation"]) -> kgen50_ph_2add.df[cs, "subpopulation"]
  as.character(kgen50_cases.df[cs, "population"]) -> kgen50_ph_2add.df[cs, "population"]
  as.character(kgen50_cases.df[cs, "gender"]) -> kgen50_ph_2add.df[cs, "gender"]
}

phenotypes.df <- rbind(kgen50_ph_2add.df, phenotypes.df)

# Rename non-changed tables
exac.df <- wecare_nfe_exac.df
kgen.df <- wecare_nfe_kgen.df

# Clean-up
rm(kgen50_vcf.df, kgen50_gt.mx, kgen50_split_var_ids, wecare_nfe_split_var_ids, 
   overlap_split_var_ids, kgen50_gt_2add.mx, wecare_nfe_genotypes.mx, var, cs, 
   wecare_nfe_phenotypes.df, kgen50_cases.df, kgen50_ph_2add.df, wecare_nfe_exac.df, 
   wecare_nfe_kgen.df, wecare_nfe_variants.df)

```

# data_summary

```{r data_summary}

ls()

dim(genotypes.mx)
class(genotypes.mx)
genotypes.mx[1:5, 1:5]

dim(phenotypes.df)
str(phenotypes.df)
phenotypes.df[1:5, 1:5]

dim(variants.df)
str(variants.df)
variants.df[1:5,1:5]

dim(kgen.df)
str(kgen.df)
kgen.df[1:5,1:5]

dim(exac.df)
str(exac.df)
exac.df[1:5,1:5]

# Check consistence of rownames in kgen50_gt.mx and kgen50_vcf.df
sum(rownames(genotypes.mx) != rownames(variants.df))
sum(rownames(genotypes.mx) != rownames(kgen.df))
sum(rownames(genotypes.mx) != rownames(exac.df))

sum(colnames(genotypes.mx) != rownames(phenotypes.df))

```

# save_data

```{r save_data}

save.image(paste(interim_data_folder, "r05_add_kgen50_Apr2017.RData", sep="/"))

```

# final_section

```{r final_section}

ls()
sessionInfo()
Sys.time()

```