---
title: "calculate_eigenvectors_wecare_only_Mar2017"
output: html_document
---

started: Alexey Larionov, 2016  
last updated: Alexey Larionov, 29Mar2017

# Summary

Eigenvectors are calculated using common variants only:  
5% < AF in the dataset < 95%

Overall, eigenvectors are calculated for 3 datasets:  

- wecare-only
- wecare-nfe (using overlap wecare and nfe)
- wecare-nfe-50kgen (using overlap wecare, nfe and kgen50)  

This script deals with wecare-only dataset.  

Requires f01_calculate_eigenvectors.R

# start_section

```{r start_section}

# Time stamp
Sys.time()

# Folders
#/analysis/mtgroup_share
#/scratch /medgen
setwd("/analysis/mtgroup_share/scripts/wecare_stat_03.17/scripts")
source_data_folder <- "/analysis/mtgroup_share/scripts/wecare_stat_03.17/source_data"
interim_data_folder <- "/analysis/mtgroup_share/scripts/wecare_stat_03.17/interim_data"

# Required libraries
library(ggplot2)
library(plotly)

```

# load_data

```{r load_data}

load(paste(interim_data_folder, "r04b_add_kgen50_Mar2017.RData", sep="/"))

```

# check_data

728 samples = 480 wecare + 198 nfe + 50 kgen50  

```{r check_data}

ls()

dim(genotypes.mx)
class(genotypes.mx)
genotypes.mx[1:5, 1:5]

dim(phenotypes.df)
str(phenotypes.df)
phenotypes.df[1:5, 1:5]

dim(variants.df)
str(variants.df)
variants.df[1:5,1:5]

dim(kgen.df)
str(kgen.df)
kgen.df[1:5,1:5]

dim(exac.df)
str(exac.df)
exac.df[1:5,1:5]

# Check consistence of rownames in kgen50_gt.mx and kgen50_vcf.df
sum(rownames(genotypes.mx) != rownames(variants.df))
sum(rownames(genotypes.mx) != rownames(kgen.df))
sum(rownames(genotypes.mx) != rownames(exac.df))

sum(colnames(genotypes.mx) != rownames(phenotypes.df))

```

# make_wecare_only_subset_of_data

keep 480 wecare cases passed filters

```{r make_wecare_only_subset_of_data}

# Remove nfe cases
wecare_cases <- as.vector(phenotypes.df[phenotypes.df$cc > -1,"wes_id"])
length(wecare_cases) # 480

wecare_genotypes.mx <- genotypes.mx[,wecare_cases]
dim(wecare_genotypes.mx) # 275,516 x 480
wecare_genotypes.mx[1:5,1:5]

wecare_phenotypes.df <- phenotypes.df[wecare_cases,]
dim(wecare_phenotypes.df) # 480 x 32
wecare_phenotypes.df[1:5,1:5]

wecare_variants.df <- variants.df
wecare_exac.df <- exac.df
wecare_kgen.df <- kgen.df

# Clean-up
rm(wecare_cases, genotypes.mx, phenotypes.df, 
   variants.df, exac.df, kgen.df)

```

# remove_variants_with_uniform_genotypes_accross_all_wecare_samples
Remove 50,141 variants: 275,516 -> 225,375

```{r remove_variants_with_uniform_genotypes_accross_all_wecare_samples}

# Check that there is no all-NA variants 
# (this is assumed for some downstream calculations)
non_NA_count.udf <- function(x){sum(!is.na(x))}
all_NA <- apply(wecare_genotypes.mx, 1, non_NA_count.udf) == 0
sum(all_NA) # 0

# Function to detect uniform numeric vector
uniform_vector.udf <- function(x){
  if(min(x, na.rm=TRUE) == max(x, na.rm=TRUE)){return(TRUE)} else {return(FALSE)}}

# Variants with uniform genotypes accross all samples 
uniform_genotypes <- apply(wecare_genotypes.mx, 1, uniform_vector.udf)
summary(uniform_genotypes)
sum(uniform_genotypes) # 50,141

# Remove variants with uniform genotypes accross all samples
wecare_genotypes.mx <- wecare_genotypes.mx[!uniform_genotypes,]
dim(wecare_genotypes.mx) # 225375    480
wecare_genotypes.mx[1:5,1:5]

wecare_variants.df <- wecare_variants.df[!uniform_genotypes,]
dim(wecare_variants.df) # 225375     24
wecare_variants.df[1:5,1:5]

wecare_kgen.df <- wecare_kgen.df[!uniform_genotypes,]
dim(wecare_kgen.df) # 225375      9
wecare_kgen.df[1:5,1:5]

wecare_exac.df <- wecare_exac.df[!uniform_genotypes,]
dim(wecare_exac.df) # 225375     48
wecare_exac.df[1:5,1:5]

# Clean-up
rm(non_NA_count.udf, all_NA, uniform_vector.udf, uniform_genotypes)

```

# recalculate_total_AFs

Used for selecting common variants for eigenvectors computation.  
Could be recalculated later after exclusion of eigenvectors outliers.  

```{r recalculate_total_AFs}

# Function to calculate AN
get_allele_number.udf <- function(x){2*sum(!is.na(x))}

# Rename AF fields in the variants table
vars_colnames <- colnames(wecare_variants.df)
"ac_raw" -> vars_colnames[ vars_colnames == "AC" ]
"an_raw" -> vars_colnames[ vars_colnames == "AN" ]
"af_raw" -> vars_colnames[ vars_colnames == "AF" ]
vars_colnames -> colnames(wecare_variants.df)

# Calculate total wecare ac, an and af
ac_all <- apply(wecare_genotypes.mx, 1, sum, na.rm=TRUE)
an_all <- apply(wecare_genotypes.mx, 1, get_allele_number.udf)
af_all <- ac_all/an_all

# Add new AFs to the variants table
wecare_variants.df <- cbind(wecare_variants.df, ac_all, an_all, af_all)

# Clean-up
rm(get_allele_number.udf, ac_all, an_all, af_all, vars_colnames)

```

# calculate_ubc_AFs

```{r calculate_ubc_AFs}

# Function to calculate AN
get_allele_number.udf <- function(x){2*sum(!is.na(x))}

# Prepare genotypes table
ubc_cases <- wecare_phenotypes.df$cc == 0 
sum(ubc_cases) # 245
ubc_genotypes.mx <- wecare_genotypes.mx[,ubc_cases]
dim(ubc_genotypes.mx)

# Calculate ubc ac, an and af
ac_ubc <- apply(ubc_genotypes.mx, 1, sum, na.rm=TRUE)
an_ubc <- apply(ubc_genotypes.mx, 1, get_allele_number.udf)
af_ubc <- ac_ubc/an_ubc

# Add updated AFs to wecare_variants.df
wecare_variants.df <- cbind(wecare_variants.df, ac_ubc, an_ubc, af_ubc)

# Clean-up
rm(get_allele_number.udf, ubc_cases, ubc_genotypes.mx, ac_ubc, an_ubc, af_ubc)

```

# calculate_cbc_AFs

```{r calculate_cbc_AFs}

# Function to calculate AN
get_allele_number.udf <- function(x){2*sum(!is.na(x))}

# Prepare genotypes table
cbc_cases <- wecare_phenotypes.df$cc == 1 
sum(cbc_cases) # 235
cbc_genotypes.mx <- wecare_genotypes.mx[,cbc_cases]
dim(cbc_genotypes.mx)

# Calculate cbc ac, an and af
ac_cbc <- apply(cbc_genotypes.mx, 1, sum, na.rm=TRUE)
an_cbc <- apply(cbc_genotypes.mx, 1, get_allele_number.udf)
af_cbc <- ac_cbc/an_cbc

# Add updated AFs to wecare_variants.df
wecare_variants.df <- cbind(wecare_variants.df, ac_cbc, an_cbc, af_cbc)

# Clean-up
rm(get_allele_number.udf, cbc_cases, cbc_genotypes.mx, ac_cbc, an_cbc, af_cbc)

```

# calculate_eigenvectors

Only common variants (AF > 0.05) will be used for eigenvectors calculation.  

Note exclusion on both sides: low- and high- AFs:  
- Low AFs remove rare variants with common allele in reference genome  
- Hight AFs remove rare variants with common allele in reference genome  

```{r calculate_eigenvectors}

# --- Make subset of data with common variants only

cbc_common_vars <- wecare_variants.df$af_cbc > 0.05 & wecare_variants.df$af_cbc < 0.95
sum(cbc_common_vars) # 44,407

ubc_common_vars <- wecare_variants.df$af_ubc > 0.05 & wecare_variants.df$af_ubc < 0.95
sum(ubc_common_vars) # 44,350

common_overlap_vars <- cbc_common_vars & ubc_common_vars
sum(common_overlap_vars) # 42,868

min(wecare_variants.df$af_all[common_overlap_vars]) # ~0.05
max(wecare_variants.df$af_all[common_overlap_vars]) # ~0.95

common_overlap_genotypes.mx <- wecare_genotypes.mx[common_overlap_vars,]
dim(common_overlap_genotypes.mx)
common_overlap_genotypes.mx[1:5,1:5]

# --- Calculate eigenvectors --- #

source("f01_calculate_eigenvectors.R")
wecare.eigen <- normalise_and_calculate_eigenvectors.udf(common_overlap_genotypes.mx)

# Clean-up
rm(cbc_common_vars, ubc_common_vars, common_overlap_vars, 
   common_overlap_genotypes.mx, normalise_and_calculate_eigenvectors.udf)

```

# plot_eigenvectors

```{r plot_eigenvectors}

# --- Prepare data for plotting --- #

wecare.eigenvectors.df <- as.data.frame(wecare.eigen$vectors) # eigenvectors in columns

# Prepare colour scale
colours <- c("UBC" = "BLUE", "CBC" = "RED")
userColourScale <- scale_colour_manual(values=colours)

# Prepare cases lables
cases_labels <- as.vector(wecare_phenotypes.df$cc)
"CBC" -> cases_labels[cases_labels==1]
"UBC" -> cases_labels[cases_labels==0]

summary(as.factor(cases_labels))

# Prepare cases IDs (for labels on onteractive plot)
cases_IDs <- as.vector(wecare_phenotypes.df$wes_id)

# make the dataframe
data2plot.df <- cbind(cases_IDs, cases_labels, wecare.eigenvectors.df[,1:5])
colnames(data2plot.df) <- c("wes_id", "group", "ev1", "ev2", "ev3", "ev4", "ev5")

# --- Plot eig1 vs eig2 --- #

g <- ggplot(data2plot.df, aes(ev1, ev2)) +
  geom_point(aes(colour=group, fill=group, text = wes_id)) + 
  labs(title="wecare common variants<br>(42,868 x 480)", x ="eigenvector1", y = "eigenvector2") +
  userColourScale

ggplotly(g)

# --- Plot eig2 vs eig3 --- #

g <- ggplot(data2plot.df, aes(ev2, ev3)) +
  geom_point(aes(colour=group, fill=group, text = wes_id)) + 
  labs(title="wecare common variants<br>(42,868 x 480)", x ="eigenvector2", y = "eigenvector3") +
  userColourScale

ggplotly(g)

# --- Plot eig3 vs eig4 --- #

g <- ggplot(data2plot.df, aes(ev3, ev4)) +
  geom_point(aes(colour=group, fill=group, text = wes_id)) + 
  labs(title="wecare common variants<br>(42,868 x 480)", x ="eigenvector3", y = "eigenvector4") +
  userColourScale

ggplotly(g)

# --- Plot eig4 vs eig5 --- #

g <- ggplot(data2plot.df, aes(ev4, ev5)) +
  geom_point(aes(colour=group, fill=group, text = wes_id)) + 
  labs(title="wecare common variants<br>(42,868 x 480)", x ="eigenvector4", y = "eigenvector5") +
  userColourScale

ggplotly(g)

# --- Clean-up --- #

rm(wecare.eigenvectors.df, g, data2plot.df, cases_IDs, cases_labels, colours, userColourScale)

```

# calculate_outliers

Explore 6 standard deviations in 5 top eigenvectors  

```{r calculate_outliers}

wecare.eigenvectors.mx <- wecare.eigen$vectors # eigenvectors in columns

ev1 <- wecare.eigenvectors.mx[,1]
ev1.positive_outliers <- ev1 > mean(ev1) + 6 * sd(ev1)
ev1.negative_outliers <- ev1 < mean(ev1) - 6 * sd(ev1)
sum(ev1.positive_outliers)
sum(ev1.negative_outliers)

wecare_phenotypes.df$wes_id[ev1.positive_outliers]
wecare_phenotypes.df$wes_id[ev1.negative_outliers]

ev2 <- wecare.eigenvectors.mx[,2]
ev2.positive_outliers <- ev2 > mean(ev2) + 6 * sd(ev2)
ev2.negative_outliers <- ev2 < mean(ev2) - 6 * sd(ev2)
sum(ev2.positive_outliers)
sum(ev2.negative_outliers)

wecare_phenotypes.df$wes_id[ev2.positive_outliers]
wecare_phenotypes.df$wes_id[ev2.negative_outliers]

ev3 <- wecare.eigenvectors.mx[,3]
ev3.positive_outliers <- ev3 > mean(ev3) + 6 * sd(ev3)
ev3.negative_outliers <- ev3 < mean(ev3) - 6 * sd(ev3)
sum(ev3.positive_outliers)
sum(ev3.negative_outliers)

wecare_phenotypes.df$wes_id[ev3.positive_outliers]
wecare_phenotypes.df$wes_id[ev3.negative_outliers]

ev4 <- wecare.eigenvectors.mx[,4]
ev4.positive_outliers <- ev4 > mean(ev4) + 6 * sd(ev4)
ev4.negative_outliers <- ev4 < mean(ev4) - 6 * sd(ev4)
sum(ev4.positive_outliers)
sum(ev4.negative_outliers)

wecare_phenotypes.df$wes_id[ev4.positive_outliers]
wecare_phenotypes.df$wes_id[ev4.negative_outliers]

ev5 <- wecare.eigenvectors.mx[,5]
ev5.positive_outliers <- ev5 > mean(ev5) + 6 * sd(ev5)
ev5.negative_outliers <- ev5 < mean(ev5) - 6 * sd(ev5)
sum(ev5.positive_outliers)
sum(ev5.negative_outliers)

wecare_phenotypes.df$wes_id[ev5.positive_outliers]
wecare_phenotypes.df$wes_id[ev5.negative_outliers]

# Clean-up
rm(wecare.eigenvectors.mx, 
   ev1, ev1.positive_outliers, ev1.negative_outliers, 
   ev2, ev2.positive_outliers, ev2.negative_outliers, 
   ev3, ev3.positive_outliers, ev3.negative_outliers,
   ev4, ev4.positive_outliers, ev4.negative_outliers,
   ev5, ev5.positive_outliers, ev5.negative_outliers)

```

# data_summary

```{r data_summary}

ls()

dim(wecare_genotypes.mx)
class(wecare_genotypes.mx)
wecare_genotypes.mx[1:5,1:5]

dim(wecare_phenotypes.df)
str(wecare_phenotypes.df)
wecare_phenotypes.df[1:5,1:5]

dim(wecare_variants.df)
colnames(wecare_variants.df)
wecare_variants.df[1:5,1:5]

dim(wecare_kgen.df)
colnames(wecare_kgen.df)
wecare_kgen.df[1:5,1:5]

dim(wecare_exac.df)
colnames(wecare_exac.df)
wecare_exac.df[1:5,1:5]

str(wecare.eigen)

sum(colnames(wecare_genotypes.mx) != rownames(wecare_phenotypes.df))
sum(rownames(wecare_genotypes.mx) != rownames(wecare_variants.df))
sum(rownames(wecare_genotypes.mx) != rownames(wecare_kgen.df))
sum(rownames(wecare_genotypes.mx) != rownames(wecare_exac.df))

```

# save_data

```{r save_data}

save.image(paste(interim_data_folder, "r05a_calculate_egenvectors_wecare_only_Mar2017.RData", sep="/"))

```

# final_section

```{r final_section}

ls()
sessionInfo()
Sys.time()

```
