---
title: "filter_by_variant_effect_Apr2017"
output: html_document
---

started: Alexey Larionov, 2016  
last updated: Alexey Larionov, 09Apr2017

# Summary
## Numbers to be corrected after decision about the eigenvector outliers exclusion!

274,037 -> 23,523 variants (10,159 genes, ~2.3 variant per gene)

### Including Loss of function variants
"splice_acceptor_variant", "splice_donor_variant", "stop_gain", "stop_lost", 
"start_lost", "frameshift_variant" : 5,007 variants

### Including dd-missense variants, likely affecting protein function
"deleterious" and "probably_damaging": 18,050 variants (of 106,983 total missenses)  
add NMD-transcripts?  

### Including ClinSig variants
"likely_pathogenic", "risk_factor", "pathogenic", "association", "protective", 
"drug_response": 842 variants

## NOT excluding variants with high AF in 1k 

These are sequence versions from Ref genome, which ALTs are common in 1k:

EUR_AF>0.5: 156 from selected  (<1%)
EUR_AF>0.9: 28 from selected

These sequences are preserved in this script.  
However, they will need special treatment in downstream analysis. 

## NOT excluding multiallelic variaints
604 multiallelic varaints in selected  
(7,178 in the whole variantset)

# start_section

```{r start_section}

# Start time
Sys.time()

# Folders
#/analysis/mtgroup_share
#/scratch/medgen
setwd("/analysis/mtgroup_share/scripts/wecare_stat_04.17/scripts")
source_data_folder <- "/analysis/mtgroup_share/scripts/wecare_stat_04.17/source_data"
interim_data_folder <- "/analysis/mtgroup_share/scripts/wecare_stat_04.17/interim_data"
results_folder <- "/analysis/mtgroup_share/scripts/wecare_stat_04.17/results"

# Load libraries
library(dplyr)
library(reshape2) # for melt()
library(ggplot2)
library(VennDiagram)

```

# load_data

```{r load_data}

load(paste(interim_data_folder, "r07_pick_egenvectors_exclude_outliers_Apr2017.RData", sep="/"))

```

# check_data

```{r check_data}

ls()

dim(wecare_nfe_genotypes.mx)
class(wecare_nfe_genotypes.mx)
wecare_nfe_genotypes.mx[1:5,1:5]

dim(wecare_nfe_phenotypes.df)
str(wecare_nfe_phenotypes.df)
wecare_nfe_phenotypes.df[1:5,1:5]

dim(wecare_nfe_variants.df)
colnames(wecare_nfe_variants.df)
wecare_nfe_variants.df[1:5,1:5]

dim(wecare_nfe_kgen.df)
colnames(wecare_nfe_kgen.df)
wecare_nfe_kgen.df[1:5,1:5]

dim(wecare_nfe_exac.df)
colnames(wecare_nfe_exac.df)
wecare_nfe_exac.df[1:5,1:5]

sum(rownames(wecare_nfe_genotypes.mx) != rownames(wecare_nfe_variants.df))
sum(rownames(wecare_nfe_genotypes.mx) != rownames(wecare_nfe_kgen.df))
sum(rownames(wecare_nfe_genotypes.mx) != rownames(wecare_nfe_exac.df))

sum(colnames(wecare_nfe_genotypes.mx) != rownames(wecare_nfe_phenotypes.df))

```

# --- explore_variants_annotations --- #

## VEP_consequences

A combination of the below consequencies can be assigned to a variant  
"incomplete_terminal_codon_variant" added to wecare only

```{r VEP_consequences}

sum(is.na(wecare_nfe_variants.df$Consequence)) # 0
a <- strsplit(as.vector(wecare_nfe_variants.df$Consequence),"&")
b <- unique(as.vector(unlist(a)))
b
rm(a,b)

```

## sift, polyphen

### Annotations seen in wecare-nfe

*SIFT:* "tolerated", "deleterious", "tolerated_low_confidence", "deleterious_low_confidence"  

*PolyPhen:* "benign", "possibly_damaging", "probably_damaging", "unknown"  

### Interpreting SIFT and PolyPhen scores (not used here)
*SIFT:* the higher score the less deleterious  
*PolyPhen:* the higher score the more damaging

```{r explore_SIFT_PolyPhen}

# SIFT
unique(as.vector(wecare_nfe_variants.df$SIFT_call))

# PolyPhen
unique(as.vector(wecare_nfe_variants.df$PolyPhen_call))

```

## clin_sig

### Annotations seen in wecare-nfe

*CLIN_SIG:* "likely_benign", "benign", "likely_pathogenic", "risk_factor", "not_provided", 
"uncertain_significance", "pathogenic", "association", "protective", "other", "drug_response"  

The same as in wecare-only. I have also seen "confers_sensitivity" in CLIN_SIG for gastric cases

```{r clin_sig}

unique(unlist(strsplit(as.vector(wecare_nfe_variants.df$CLIN_SIG), split="\\&")))

```

## count_variants_frequent_in_kgen_and_exac

2,350 rare sequence versions are present into reference genome (b37)  
~1% of all variants in the datset

Tehy will be preserbved; however they should be treated specially 
in downstream analyses because "variants" called in these sites represent 
sequence versions common in humans, despite the misleading "low" MAFs in some databases. 

```{r count_variants_frequent_in_kgen_and_exac}

x <- wecare_nfe_kgen.df  %>% filter(kgen.EUR_AF > 0.9) %>% select(SplitVarID) # 2,350
y <- wecare_nfe_exac.df  %>% filter(exac_non_TCGA.AF > 0.9) %>% select(SplitVarID) # 1,728

x <- as.character(x[,1])
y <- as.character(y[,1])
z <- intersect(x,y)

length(x) # 2,369
length(y) # 1,744
length(z) # 1,627

rm(x,y,z)

```

## count_multiallelic_varaints

7,178 multiallelic varaints

```{r count_multiallelic_varaints}

summary(wecare_nfe_variants.df$Multiallelic) # 7,178

```

# --- Make indices --- #

## make_lof_index

```{r make_lof_index}

sum(is.na(wecare_nfe_variants.df$Consequence)) # 0

# --- Splice acceptor variants (sav) --- #

splice_acceptor_variant <- grepl( "splice_acceptor_variant", wecare_nfe_variants.df$Consequence)
sum(splice_acceptor_variant) # 423

# mean count of alleles per case in wecare
sav_wecare_nfe.mx <- wecare_nfe_genotypes.mx[splice_acceptor_variant,]
sav_wecare_nfe.counts <- apply(sav_wecare_nfe.mx, 2, sum, na.rm=TRUE)
sav_wecare_nfe.mean <- mean(sav_wecare_nfe.counts, na.rm=TRUE) # ~22
rm(sav_wecare_nfe.mx, sav_wecare_nfe.counts)

# --- Splice donor variants (sdv) --- #

splice_donor_variant <- grepl( "splice_donor_variant", wecare_nfe_variants.df$Consequence)
sum(splice_donor_variant) # 625

# mean count of alleles per case in wecare
sdv_wecare_nfe.mx <- wecare_nfe_genotypes.mx[splice_donor_variant,]
sdv_wecare_nfe.counts <- apply(sdv_wecare_nfe.mx, 2, sum, na.rm=TRUE)
sdv_wecare_nfe.mean <- mean(sdv_wecare_nfe.counts, na.rm=TRUE) # ~16
rm(sdv_wecare_nfe.mx, sdv_wecare_nfe.counts)

# --- Stop gains --- #

stop_gain <- grepl( "stop_gain", wecare_nfe_variants.df$Consequence)
sum(stop_gain) # 2,189

# mean count of alleles per case in wecare
stogv_wecare_nfe.mx <- wecare_nfe_genotypes.mx[stop_gain,]
stogv_wecare_nfe.counts <- apply(stogv_wecare_nfe.mx, 2, sum, na.rm=TRUE)
stogv_wecare_nfe.mean <- mean(stogv_wecare_nfe.counts, na.rm=TRUE) # ~50
rm(stogv_wecare_nfe.mx, stogv_wecare_nfe.counts)

# --- Stop losts --- #

# Interestingly, there are much less of stop_losts than stop_gains: 79 and 2,189 respectively. 
# This looks contraintuitive, suggesting that stop_gains are better tolerated than stop losses. 
# In other words: stop_losses may be under stronger evolutionary control than stop_gains??

# An alternative explanation coulod be that stop losts can only happen on stop codons; 
# while stop gain can happen in "any"" codon within the gene: so stop gains have much more chances to happen. 

stop_lost <- grepl( "stop_lost", wecare_nfe_variants.df$Consequence)
sum(stop_lost) # 79

# mean count of alleles per case in wecare
stolv_wecare_nfe.mx <- wecare_nfe_genotypes.mx[stop_lost,]
stolv_wecare_nfe.counts <- apply(stolv_wecare_nfe.mx, 2, sum, na.rm=TRUE)
stolv_wecare_nfe.mean <- mean(stolv_wecare_nfe.counts, na.rm=TRUE) # ~15
rm(stolv_wecare_nfe.mx, stolv_wecare_nfe.counts)

# --- Start losts --- #

# Functional significance of start losses may be questioned: a heterozigous start loss may
# be easily compemsated by a mere expression of retained allele?? 

# However, there is not much of them... Either this is an evolutionary control or just "low base" effect (as above for the stop_losses)?

# The total number of start losses is small, so they should not influence the analysis too much anyway ... 

# Why there is no such consequence as "Start gain" ?

start_lost <- grepl( "start_lost", wecare_nfe_variants.df$Consequence)
sum(start_lost) # 113

# mean count of alleles per case in wecare
stalv_wecare_nfe.mx <- wecare_nfe_genotypes.mx[start_lost,]
stalv_wecare_nfe.counts <- apply(stalv_wecare_nfe.mx, 2, sum, na.rm=TRUE)
stalv_wecare_nfe.mean <- mean(stalv_wecare_nfe.counts, na.rm=TRUE) # ~7
rm(stalv_wecare_nfe.mx, stalv_wecare_nfe.counts)

# --- Frameshifts --- #

frameshift_variant <- grepl( "frameshift_variant", wecare_nfe_variants.df$Consequence)
sum(frameshift_variant) # 1,623

# mean count of alleles per case in wecare
fsv_wecare_nfe.mx <- wecare_nfe_genotypes.mx[frameshift_variant,]
fsv_wecare_nfe.counts <- apply(fsv_wecare_nfe.mx, 2, sum, na.rm=TRUE)
fsv_wecare_nfe.mean <- mean(fsv_wecare_nfe.counts, na.rm=TRUE) # ~124
rm(fsv_wecare_nfe.mx, fsv_wecare_nfe.counts)

# --- all LoF-s --- #

lof <- splice_acceptor_variant | splice_donor_variant | stop_gain | stop_lost | start_lost | frameshift_variant

sum(lof) # 5,007

# mean count of alleles per case in wecare
lof_wecare_nfe.mx <- wecare_nfe_genotypes.mx[lof,]
lof_wecare_nfe.counts <- apply(lof_wecare_nfe.mx, 2, sum, na.rm=TRUE)
lof_wecare_nfe.mean <- mean(lof_wecare_nfe.counts, na.rm=TRUE) # ~228
rm(lof_wecare_nfe.mx, lof_wecare_nfe.counts)

# ------- Plot average counts of lof-s per individual ------- #

# Summary table with average numbers of lof-s per individual
lof_summary_counts.df <- as.data.frame(cbind(
  lof_type=c("splice_acceptor", "splice_donior", "stop_gain", "stop_lost", "start_lost", "frameshift", "all_LoFs"),
  lof_count=c(sav_wecare_nfe.mean, sdv_wecare_nfe.mean, stogv_wecare_nfe.mean, stolv_wecare_nfe.mean, stalv_wecare_nfe.mean, fsv_wecare_nfe.mean, lof_wecare_nfe.mean)))

lof_summary_counts.df$lof_count <- as.numeric(as.vector(lof_summary_counts.df$lof_count))
lof_summary_counts.df <- lof_summary_counts.df %>% arrange(desc(lof_count))

# Order of levels defines order of bars on the plot
lof_summary_counts.df$lof_type <- factor(lof_summary_counts.df$lof_type, 
  levels=c("all_LoFs", "frameshift", "stop_gain", "splice_acceptor", "splice_donior", "stop_lost", "start_lost"))

# Plot
ggplot(lof_summary_counts.df, aes(lof_type, lof_count)) +
  geom_bar(stat="identity", fill=c("red",rep("blue",6)), alpha=0.5) +
  geom_text(aes(label=round(lof_count,1)), vjust=-1, size=5)+
  labs(x = "", y = "") + 
  ylim(0,250) + 
  theme(axis.text.x = element_text(hjust=1, size=18, angle=45)) +
  ggtitle("Average counts of LoF variants per individual (wecare-nfe)")
  
# Clean-up
rm(splice_acceptor_variant, splice_donor_variant, stop_gain, frameshift_variant, stop_lost,
   start_lost, sav_wecare_nfe.mean, sdv_wecare_nfe.mean, stogv_wecare_nfe.mean, stolv_wecare_nfe.mean,
   stalv_wecare_nfe.mean, fsv_wecare_nfe.mean, lof_summary_counts.df)

```

## make_missense_index

Selecting dd-missense variants:  
both: [ deleterious by SIFT ] AND [ probably_damaging by PolyPhen ]  

```{r make_missense_index}

missense_all <- grepl( "missense_variant", wecare_nfe_variants.df$Consequence)
sum(missense_all) # 106,983

summary(wecare_nfe_variants.df$SIFT_call) 
# note many NAs; these are when SIFT is irrelevant (e.g. non-coding)?
deleterious <- wecare_nfe_variants.df$SIFT_call == "deleterious"
summary(deleterious)
FALSE -> deleterious[is.na(deleterious)]
summary(deleterious) # 37,684

summary(wecare_nfe_variants.df$PolyPhen_call) 
# note many NAs; these are when PolyPhen is irrelevant (e.g. non-coding)?
probably_damaging <- wecare_nfe_variants.df$PolyPhen_call == "probably_damaging"
summary(probably_damaging)
FALSE -> probably_damaging[is.na(probably_damaging)]
summary(probably_damaging) # 24,570

missense <- deleterious & probably_damaging
sum(missense) # 18,050

# Plot venn diagram of selected variants by types
grid.newpage()
draw.pairwise.venn(
  area1=sum(deleterious), 
  area2=sum(probably_damaging), 
  cross.area=sum(missense), 
  category=c(
    paste("SIFT\ndeleterious\n", sum(deleterious)),
    paste("PolyPhen\nprobably_damaging\n", sum(probably_damaging))),
  fill=c("red", "blue"), 
  alpha = c(0.3, 0.3),
  cex=2, cat.fontface=4, fontfamily=3)

# mean count of missenses per case in wecare
missense_wecare_nfe.mx <- wecare_nfe_genotypes.mx[missense,]
missense_wecare_nfe.counts <- apply(missense_wecare_nfe.mx, 2, sum, na.rm=TRUE)
missense_wecare_nfe.mean <- mean(missense_wecare_nfe.counts, na.rm=TRUE)
missense_wecare_nfe.mean # ~375

# Clean-up
rm(missense_all, deleterious, probably_damaging, missense_wecare_nfe.mx, missense_wecare_nfe.counts)

```

## make_clinsig_index

ClinSig version as reported by Ensembl v87-grch37_vep_cache (Jan2017, VEP script)
ClinVar may become a better variant-annotation database, which may be used in future. 

All *CLIN_SIG* annotations found in the file: "likely_benign", "benign", "likely_pathogenic", "risk_factor", "not_provided", "uncertain_significance", "pathogenic", "association", "protective", "other", "drug_response"  

Note sums w/o na.rm=TRUE.

```{r make_clinsig_index}

selected_clinsig_terms <- c("likely_pathogenic", "risk_factor", "pathogenic", "association", "protective", "drug_response")

clinsig <- wecare_nfe_variants.df$CLIN_SIG %in% selected_clinsig_terms
sum(clinsig) # 842

x.df <- as.data.frame(table(as.vector(wecare_nfe_variants.df[clinsig,"CLIN_SIG"])))
c("type","count") -> colnames(x.df)
x.df$type <- factor(x.df$type, 
  levels=c("pathogenic", "risk_factor", "likely_pathogenic", "drug_response", "protective", "association"))

ggplot(x.df, aes(type, count)) +
  geom_bar(stat="identity", fill="blue", alpha=0.5) +
  geom_text(aes(label=round(count,1)), vjust=-1, size=5)+
  labs(x = "", y = "") + 
  ylim(0,700) + 
  theme(axis.text.x = element_text(hjust=1, size=18, angle=45)) +
  ggtitle("Counts of selected ClinSig terms in wecare-nfe variantset")

# mean count of alleles per case in wecare
clinsig_wecare_nfe.mx <- wecare_nfe_genotypes.mx[clinsig,]
clinsig_wecare_nfe.counts <- apply(clinsig_wecare_nfe.mx, 2, sum, na.rm=TRUE)
clinsig_wecare_nfe.mean <- mean(clinsig_wecare_nfe.counts, na.rm=TRUE) 
clinsig_wecare_nfe.mean # 78.5

# Clean-up
rm(selected_clinsig_terms, clinsig_wecare_nfe.mx, clinsig_wecare_nfe.counts, x.df)

```

# --- explore_selected --- #

## selected_types_per_individual_and_per_variantset

```{r selected_types_per_individual_and_per_variantset}

# Count combinations of selectred categories
lof_missense <- lof & missense
clinsig_missense <- clinsig & missense
lof_clinsig <- lof & clinsig
lof_clinsig_missense <- lof_clinsig & missense

summary(lof_missense)
summary(clinsig_missense)
summary(lof_clinsig)
summary(lof_clinsig_missense)

selected_variants <- lof | clinsig | missense

summary(selected_variants) # 23,523

# --- Plot mean counts of variant types per individual --- #

# Summary table with average numbers of protein-affecting per individual
summary_counts.df <- as.data.frame(cbind(
  type=c("Missense", "LoF", "ClinSig"),
  count=c(missense_wecare_nfe.mean, lof_wecare_nfe.mean, clinsig_wecare_nfe.mean)))

# Convert numbers to numeric vector
summary_counts.df$count <- as.numeric(as.vector(summary_counts.df$count))

# Order of levels defines order of bars on the plot
summary_counts.df$type <- factor(summary_counts.df$type, 
  levels=c("Missense", "LoF", "ClinSig"))

# Plot bar diagramm of avg. count of selected types per individual
ggplot(summary_counts.df, aes(type, count)) +
  geom_bar(stat="identity", fill="blue", alpha=0.5) + 
  geom_text(aes(label=round(count,1)), vjust=-1, size=5)+
  ylim(0, 420) +
  labs(x = "", y = "") + 
  theme(axis.text.x = element_text(hjust=1, size=18, angle=45)) +
  ggtitle("Average counts of selected variants per individual (wecare-nfe)")

# Plot venn diagram of selected variants by types
grid.newpage()
draw.triple.venn(
  area1=sum(lof), 
  area2=sum(clinsig), 
  area3=sum(missense), 
  n12=sum(lof_clinsig), 
  n23=sum(clinsig_missense), 
  n13=sum(lof_missense),
  n123=sum(lof_clinsig_missense),
  category=c(
    paste("lof\n", sum(lof)),
    paste("clinsig\n", sum(clinsig)),
    paste("missense\n", sum(missense))),
  fill = c("red", "green", "blue"), 
  alpha = c(0.3, 0.3, 0.3),
  cex=2, cat.fontface=4, fontfamily=3)

rm(missense_wecare_nfe.mean, lof_wecare_nfe.mean, clinsig_wecare_nfe.mean, 
   lof, clinsig, missense, clinsig_missense, lof_clinsig, lof_missense, lof_clinsig_missense, 
   summary_counts.df)

```

# keep_selected_variants_only

274,037 -> 23,364

```{r keep_selected_variants_only}

wecare_nfe_variants.df <- wecare_nfe_variants.df[selected_variants,]
wecare_nfe_kgen.df <- wecare_nfe_kgen.df[selected_variants,]
wecare_nfe_exac.df <- wecare_nfe_exac.df[selected_variants,]
wecare_nfe_genotypes.mx <- wecare_nfe_genotypes.mx[selected_variants,]


dim(wecare_nfe_variants.df)
dim(wecare_nfe_kgen.df)
dim(wecare_nfe_exac.df)
dim(wecare_nfe_genotypes.mx)

# Clean-up
rm(selected_variants)
  
```

# check_frequent_in_Kgen

These variants are preserved;
However, they will require special treatment in downstream analysis

```{r check_frequent_in_Kgen}

x <- wecare_nfe_kgen.df$kgen.EUR_AF > 0.9
FALSE -> x[is.na(x)]
x <- wecare_nfe_kgen.df[x, "SplitVarID"]
x <- as.character(x)
length(x) # 28 variants with AF > 0.9
wecare_nfe_variants.df[unlist(x),c("SplitVarID","SYMBOL")]

y <- wecare_nfe_kgen.df$kgen.EUR_AF > 0.5
sum(y, na.rm=TRUE) # 156 variants with AF > 0.5

rm(x,y)

```

# count_selected_multiallelic_varaints

604 multiallelics in selected varaints

```{r count_selected_multiallelic_varaints}

summary(wecare_nfe_variants.df$Multiallelic) # 604

```

# count_variants_per_gene

```{r count_variants_per_gene}

# Get number of genes
length(unique(wecare_nfe_variants.df$SYMBOL)) # 10,159

# Look at the top genes
x <- sort(table(as.vector(wecare_nfe_variants.df$SYMBOL)), decreasing = TRUE)
x[1:200]

# Save table of counts
y <- cbind(names(x),x)
c("gene","count") -> colnames(y)
write.table(y, file=paste(results_folder, "wecare_nfe_variants_per_gene.txt", sep="/"), quote = FALSE, row.names = FALSE)

# Plot histogram
z <- wecare_nfe_variants.df %>%  group_by(SYMBOL) %>% summarise(n())
"vars_count" -> colnames(z)[2]
hist(z$vars_count, main="Histogram for numbers of protein-affecting variants per gene (wecare-nfe)", xlab="Variants per gene", ylab="Number of genes", ylim=c(0,8000), breaks=36, labels = TRUE)
mean(z$vars_count) # 2.315484

# Clean-up
rm(x,y,z)

```

# explore_AFs_in_selected_variants

```{r explore_AFs_in_selected_variants}

mean(wecare_nfe_variants.df$af_all) # ~1.5%

hist(wecare_nfe_variants.df$af_all, main="AFs of selected variants")
hist(wecare_nfe_variants.df[wecare_nfe_variants.df$af_all<0.01,"af_all"],
     ylim=c(0,20000), labels = TRUE, main="zoom to <1%")

```

# calculate_aggregated_counts_per_gene

```{r calculate_aggregated_counts_per_gene}

# Get list of genes
genes <- unique(as.character(as.vector(wecare_nfe_variants.df$SYMBOL)))

# Make matrix for aggregated counts in wecare-nfe
genes.mx <- matrix(ncol=ncol(wecare_nfe_genotypes.mx),nrow=0)
colnames(genes.mx) <- colnames(wecare_nfe_genotypes.mx)

# Invert genotypes with total AF > 0.5
genotypes_inv.mx <- wecare_nfe_genotypes.mx
for(var in 1:nrow(wecare_nfe_variants.df)){
  
  if(wecare_nfe_variants.df[var,"af_all"] > 0.5){
    genotype <- wecare_nfe_genotypes.mx[var,]
    inv_genotype <- genotype
    0 -> inv_genotype[genotype == 2]
    2 -> inv_genotype[genotype == 0]
    genotypes_inv.mx[var,] <- inv_genotype
    rm(genotype, inv_genotype)
  }
}

# Calculate counts per gene 
for(gene in genes){
  
  vars <- wecare_nfe_variants.df$SYMBOL == gene
  
  if(sum(vars)==1){
    counts <- genotypes_inv.mx[vars,] # preserves NA
  }else{
    counts <- apply(genotypes_inv.mx[vars,], 2, sum, na.rm=TRUE)
  }
  
  genes.mx <- rbind(genes.mx, counts)
  rm(vars,counts)
  
}

dim(genes.mx)
counts_per_gene <- apply(genes.mx, 1, sum,na.rm=TRUE)
genes -> names(counts_per_gene)

wecare_nfe_genes.df <- data.frame(genes, counts_per_gene)
wecare_nfe_genes.df$genes <- as.character(as.vector(wecare_nfe_genes.df$genes))
rownames(wecare_nfe_genes.df) <- genes
str(wecare_nfe_genes.df)
wecare_nfe_genes.df[wecare_nfe_genes.df$genes=="ATM",]
wecare_nfe_genes.df[wecare_nfe_genes.df$genes=="CHEK2",]

# Genes with frequent protein-affecting variants
x <- counts_per_gene[order(-counts_per_gene)]
x[1:100]

hist(counts_per_gene)
hist(counts_per_gene[counts_per_gene<31], breaks=c(0:30), 
     labels = TRUE, ylim=c(0,4000), main="zoom to <30 (AF~5%)")$counts

# Counts >30 (~5%) - too common to be pathogenic
sum(counts_per_gene > 30) # 1,386 - include anyway ?

# Counts <3 (~0.5%) - too rare for detection
sum(counts_per_gene < 3) # 4,835

# Genes that may be analysed
sum(counts_per_gene >= 3 & counts_per_gene <= 30) # 3,938

# Clean-up
rm(genes.mx, genes, gene, genotypes_inv.mx, counts_per_gene, var, x)

```

# data_summary

```{r data_summary}

dim(wecare_nfe_genotypes.mx)
class(wecare_nfe_genotypes.mx)
wecare_nfe_genotypes.mx[1:5,1:5]

dim(wecare_nfe_genes.df)
class(wecare_nfe_genes.df)
wecare_nfe_genes.df[1:5,]

dim(wecare_nfe_kgen.df)
colnames(wecare_nfe_kgen.df)
wecare_nfe_kgen.df[1:5,1:5]

dim(wecare_nfe_exac.df)
colnames(wecare_nfe_exac.df)
wecare_nfe_exac.df[1:5,1:5]

dim(wecare_nfe_variants.df)
str(wecare_nfe_variants.df)
wecare_nfe_variants.df[1:5,1:5]

dim(wecare_nfe_phenotypes.df)
str(wecare_nfe_phenotypes.df)
wecare_nfe_phenotypes.df[1:5,1:5]

# Check consistency of rownames and colnames
sum(colnames(wecare_nfe_genotypes.mx) != rownames(wecare_nfe_phenotypes.df))

sum(rownames(wecare_nfe_genotypes.mx) != rownames(wecare_nfe_kgen.df))
sum(rownames(wecare_nfe_genotypes.mx) != rownames(wecare_nfe_exac.df))
sum(rownames(wecare_nfe_genotypes.mx) != rownames(wecare_nfe_variants.df))

```

# save_data

```{r save_data}

save.image(paste(interim_data_folder, "s08_filter_by_variant_effect_Apr2017.RData", sep="/"))

```

# final_section

```{r final_section}

ls()
sessionInfo()
Sys.time()

```
