---
title: "nfe vs wecare: glm QQ-plots"
output: html_document
---

started: Alexey Larionov, 25Mar2017  
last updated: Alexey Larionov, 01Apr2017  

# Summary

Make QQ plot(s) for nfe vs wecare to see whether there is an inflation 
(recommended by DC during the Skype meeting of 14Mar2017)

Use genes with > 2 variants in wecare-nfe
23,523 -> 17,482 variants
10,159 -> 5,324 genes

Before aggregation:
- Invert genotypes for variants with total AF > 0.5
- Weight by total AF beta(0,25)

Aggregate per gene by summarising selected inverted weighted variants

Make QQ-plot by using simple logit regression on the aggregated data 
- add 3 top eigenvectors to the model

Todo: add an option for calculations w/o weighting (uniform weight = 1)

# start_section

```{r start_section}

# Start time
Sys.time()

# Folders
#/analysis/mtgroup_share
#/scratch /medgen
setwd("/analysis/mtgroup_share/scripts/wecare_stat_03.17/scripts")
source_data_folder <- "/analysis/mtgroup_share/scripts/wecare_stat_03.17/source_data"
interim_data_folder <- "/analysis/mtgroup_share/scripts/wecare_stat_03.17/interim_data"
results_folder <- "/analysis/mtgroup_share/scripts/wecare_stat_03.17/results"

```

# load_data

```{r load_data}

# Load data
load(paste(interim_data_folder, "s07_filter_by_variant_effect_Mar2017.RData", sep="/"))

```

# check_data

```{r check_data}

dim(wecare_nfe_genotypes.mx)
class(wecare_nfe_genotypes.mx)
wecare_nfe_genotypes.mx[1:5,1:5]

dim(wecare_nfe_genes.df)
class(wecare_nfe_genes.df)
wecare_nfe_genes.df[1:5,]

dim(wecare_nfe_kgen.df)
colnames(wecare_nfe_kgen.df)
wecare_nfe_kgen.df[1:5,1:5]

dim(wecare_nfe_exac.df)
colnames(wecare_nfe_exac.df)
wecare_nfe_exac.df[1:5,1:5]

dim(wecare_nfe_variants.df)
str(wecare_nfe_variants.df)
wecare_nfe_variants.df[1:5,1:5]

dim(wecare_nfe_phenotypes.df)
str(wecare_nfe_phenotypes.df)
wecare_nfe_phenotypes.df[1:5,1:5]

# Check consistency of rownames and colnames
sum(colnames(wecare_nfe_genotypes.mx) != rownames(wecare_nfe_phenotypes.df))

sum(rownames(wecare_nfe_genotypes.mx) != rownames(wecare_nfe_kgen.df))
sum(rownames(wecare_nfe_genotypes.mx) != rownames(wecare_nfe_exac.df))
sum(rownames(wecare_nfe_genotypes.mx) != rownames(wecare_nfe_variants.df))

```

# select_variants

Should I use common - or at least present - in both?

Use genes with > 2 variants in wecare-nfe
23,523 -> 17,482 variants
10,159 -> 5,324 genes

```{r select_variants}

# --- Select variants for genes with > 2 aggregated events per dataset --- #

sel_genes <- wecare_nfe_genes.df[wecare_nfe_genes.df$counts_per_gene >= 3,"genes"]
length(sel_genes) # 5,324
sel_genes[1:5]

sel_vars <- wecare_nfe_variants.df[wecare_nfe_variants.df$SYMBOL %in% sel_genes, "SplitVarID"]
sel_vars <- as.character(as.vector(sel_vars))
length(sel_vars) # 17,842
sel_vars[1:5]

genotypes_sel.mx <- wecare_nfe_genotypes.mx[sel_vars,]
variants_sel.df <- wecare_nfe_variants.df[sel_vars,]

dim(genotypes_sel.mx)
dim(variants_sel.df)

```

# invert_genotypes

Invert genotypes for variants with total AF > 0.5  

```{r invert_genotypes}

# Prepare tables for output
variants_sel_inv.df <- cbind(variants_sel.df, 
                     inverted=rep(FALSE,nrow(variants_sel.df)),
                     ac_inv=variants_sel.df$ac_all,
                     an_inv=variants_sel.df$an_all,
                     af_inv=variants_sel.df$af_all)

genotypes_sel_inv.mx <- genotypes_sel.mx

# For each variant
for(var in sel_vars){
  
  # If total AF > 0.5
  if(variants_sel_inv.df[var,"af_all"] > 0.5){
    
    # Flag variant
    TRUE -> variants_sel_inv.df[var, "inverted"]
    
    # Invert genotypes 
    genotypes <- genotypes_sel_inv.mx[var,]
    genotypes_inv <- genotypes
    0 -> genotypes_inv[genotypes == 2]
    2 -> genotypes_inv[genotypes == 0]
    genotypes_inv -> genotypes_sel_inv.mx[var,]

    # Calculate inverted AFs
    ac_inv <- sum(genotypes_inv, na.rm=TRUE)
    an_inv <- 2*sum(!is.na(genotypes_inv))
    af_inv <- ac_inv/an_inv

    # Add inverted AFs to the variants data.frame
    variants_sel_inv.df[var, c("ac_inv", "an_inv", "af_inv")] <- c(ac_inv, an_inv, af_inv)
    
    # Clean-up
    rm(genotypes, genotypes_inv, ac_inv, an_inv, af_inv)
    
  }
}

# Number of invered variants
sum(variants_sel_inv.df$inverted) # 171

# Some effects of inversion
max(variants_sel_inv.df$af_all) # ~0.99
max(variants_sel_inv.df$af_inv) # 0.5

max(apply(genotypes_sel.mx,1,sum,na.rm=TRUE)) # 1,345
max(apply(genotypes_sel_inv.mx,1,sum,na.rm=TRUE)) # 666

# clean-up
rm(var, genotypes_sel.mx, variants_sel.df)

```

# impute_missed_data

During the aggregation we ned to handle NA values.  
Ideally, we could do imputation based on the haplotypes, taking into account ethnisity etc.  
Practically this imputation would be a major time investment (likely un-justifyed becuse of the rarity of our variants).  

## NA handling in SKAT
SKAT does not use the haplotype-based imputation.  
Instead SKAT allows to pick from 3 strategies to handle NAs  
(see impute.method in SKAT function help):

**bestguess**  
The most likely genotype under HW equilibrium.  
This is:  
*REF-REF* for ALT-AF < 1/3  
*REF-ALT* for ALT-AF 1/3 to 2/3  
*ALT-ALT* for ALT-AF >2/3  
For rare variants in our case this is mainly REF-REF  
(i.e. 0 in additive scoring).  

**fixed**  
2*ALT-AF, in rare variants it is close to 0  

**random**  
Random pick from binomial(2,p)  

## NA handling in this script
I use method equivalent to the "fixed" approach in SKAT:  
substitute NA with duplicated allelic frequency. 

In fact, this would be ~0 (very close to the "bestguess" for most variants)  
because of the low allelic frequencies.  Considering the callrate > 90%,  
the imputation should not have major effect either.  

```{r impute_missed_data}

# Prepare matrix for output
genotypes_sel_inv_imp.mx <- genotypes_sel_inv.mx

# Fraction of missed data
sum(is.na(genotypes_sel_inv.mx))/(nrow(genotypes_sel_inv.mx)*ncol(genotypes_sel_inv.mx))

# Impute missed data
for(var in sel_vars){

    # Get AF (corrected for the inverted variants)
    af <- variants_sel_inv.df[var,"af_inv"]
    
    # Impute missed genotypes
    2*af -> genotypes_sel_inv_imp.mx[var, is.na(genotypes_sel_inv_imp.mx[var,])]

    # Clean-up
    rm(af)

}

sum(is.na(genotypes_sel_inv_imp.mx))

# Clean-up
rm(genotypes_sel_inv.mx, var)

```

# weight_variants

Weight by total AF using beta(0,25) function (default weighting in SKAT)  

```{r weight_variants}

# Explore beta(1,25) weights
x <- c(0:100)/100
y <- dbeta(x, 1, 25)
plot(x, y, type="l", lty=1, xlab="AF", ylab="Weight", 
     xaxt="n", yaxt="n", bty='n', main = "Weighting by beta(1,25)")
axis(side=1, pos=0, at=c(0, 0.125, 0.25, 0.5, 1))
axis(side=2, pos=0, at=c(0, 5, 10, 15, 20, 25))
axis(side=2, pos=0, at=c(1)) # added separately to avoid hiding (too close to 0 to show)

abline(v=0.5, col="red", lty = 2)
abline(h=1, col="blue", lty = 2)
abline(v=0.125, col="blue", lty = 2)
abline(v=0.25, col="blue", lty = 2)

dbeta(0, 1, 25) # 25
dbeta(0.125, 1, 25) # ~1
dbeta(0.25, 1, 25) # ~0
dbeta(0.5, 1, 25) # 0
dbeta(1, 1, 25) # 0
rm(x,y)

# Prepare tables for output
variants_sel_inv_wt.df <- variants_sel_inv.df
variants_sel_inv_wt.df <- cbind(variants_sel_inv_wt.df, weight=1)

genotypes_sel_inv_imp_wt.mx <- genotypes_sel_inv_imp.mx

# Apply weighting
for(var in sel_vars){
  
  # Get AF (recalculated for inverted variants)
  af <- variants_sel_inv_wt.df[var, "af_inv"]
  
  # Calculate weight
  weight <- dbeta(af, 1, 25)
  
  # Write weight to variants table
  variants_sel_inv_wt.df[var,"weight"] <- weight
  
  # Apply weights to genotypes table
  genotypes_sel_inv_imp_wt.mx[var,] <- genotypes_sel_inv_imp_wt.mx[var,] * weight
  
  # Clean-up
  rm(af, weight)

}

# --- Explore effects of weighting --- #

# AFs
variants_sel_inv_wt.df$af_inv[6:12]

# Weights
round(variants_sel_inv_wt.df$weight[6:12],1)

# Genotypes
genotypes_sel_inv_imp.mx[6:12, 15:20]
round(genotypes_sel_inv_imp_wt.mx[6:12, 15:20], 1)

# clean-up
rm(var, genotypes_sel_inv_imp.mx, variants_sel_inv.df)

```

# aggregate_variants_per_gene

Aggregate selected, inverted, imputed, weighted variants per gene. 

```{r aggregate_variants_per_gene}

# Prepare table for the result
genes.mx <- matrix(nrow=0, ncol=ncol(genotypes_sel_inv_imp_wt.mx))
colnames(genes.mx) <- colnames(genotypes_sel_inv_imp_wt.mx)

# Aggregate (substitute NAs to 2*AF*Weight)
for(gene in sel_genes){
  
  #gene <- "ATM" # 16 vars
  #gene <- "CCNL2" # 1 var
  
  # Gene's variants
  vars <- as.vector(variants_sel_inv_wt.df[variants_sel_inv_wt.df$SYMBOL == gene, "SplitVarID"])
  
  # For a single variant per gene
  if(length(vars) == 1){
    
    # Get genotypes vector
    genotypes <- genotypes_sel_inv_imp_wt.mx[vars,]
    
    # Add to output matrix (no need in aggregation)
    aggregated_genotypes <- genotypes
    
    # Clean-up
    rm(genotypes)
      
  }else{ # for multiple variants per gene

    # Get genotypes matrix
    genotypes.mx <- genotypes_sel_inv_imp_wt.mx[vars,]
    
    # Aggregate
    aggregated_genotypes <- apply(genotypes.mx,2,sum)
    
    # Clean-up
    rm(genotypes.mx)
  }
  
  # Append
  genes.mx <- rbind(genes.mx, aggregated_genotypes)
  
  # Clean-up
  rm(aggregated_genotypes, vars)
}

# Add rownames
sel_genes -> rownames(genes.mx)

dim(genes.mx)

# Clean-up
rm(gene, genotypes_sel_inv_imp_wt.mx, variants_sel_inv_wt.df, sel_vars)

```

# prepare_data_for_analysis

No missed data in eigenvectors, outcomes or genotypes  

```{r prepare_data_for_analysis}

# Eigenvectors
E_kgen50 <- as.matrix(wecare_nfe_phenotypes.df[,c("eig1_kgen50","eig2_kgen50","eig3_kgen50")])
E_nfe <- as.matrix(wecare_nfe_phenotypes.df[,c("eig1_nfe","eig2_nfe","eig3_nfe")])
E_kgen50[1:5,]
E_nfe[1:5,]
sum(is.na(E_kgen50)) # 0
sum(is.na(E_nfe)) # 0

# Outcomes
Y <- ifelse(wecare_nfe_phenotypes.df$cc==-1, 0, 1)
sum(is.na(Y)) # 0
sum(Y) # 478
sum(!Y) # 198

# Genotypes: transpose
G <- t(genes.mx)
dim(G)
G[1:5,1:5]
sum(is.na(G)) # 0

# Clean-up
rm(genes.mx)

```

# logit_regression

Reccurent warnings:  
Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred  

```{r logit_regression, warning=FALSE}

# Write header to output matrix 
# (do not use data frame here: it will play with rbind!)
result.mx <- matrix(ncol=7, nrow=0) 
colnames(result.mx) <- c("gene", "p_glm_kgen50", "p_anova_lrt_kgen50", "p_lrt_kgen50", "p_glm_nfe", "p_anova_lrt_nfe", "p_lrt_nfe")

# For each gene
for(gene in sel_genes){
  
  # Get vector of genotypes  
  X <- G[,gene]
  
  # --- Calculate logit regression models with kgen50 eigenvectors --- #
  
  regE_kgen50 <- glm(Y ~ E_kgen50, family=binomial) # Null model
  regXE_kgen50 <- glm(Y ~ X + E_kgen50, family=binomial) # Complete model

  # Wald (?) test
  p_glm_kgen50 <- summary(regXE_kgen50)$coef["X", "Pr(>|z|)"]
  
  # Anova-LRT
  anova_lrt_kgen50 <- anova(regXE_kgen50, test="LRT")
  p_anova_lrt_kgen50 <- anova_lrt_kgen50["X", "Pr(>Chi)"]
  
  # LRT
  chi_stat_lrt_kgen50 = 2*(logLik(regXE_kgen50) - logLik(regE_kgen50))
  p_lrt_kgen50 = 1-pchisq(chi_stat_lrt_kgen50, df=1) # df??
  
  # Clean-up
  rm(regE_kgen50, regXE_kgen50, anova_lrt_kgen50, chi_stat_lrt_kgen50)
  
  # --- Calculate logit regression models with nfe eigenvectors --- #
  
  regE_nfe <- glm(Y ~ E_nfe, family=binomial) # Null model
  regXE_nfe <- glm(Y ~ X + E_nfe, family=binomial) # Complete model

  # Wald (?) test
  p_glm_nfe <- summary(regXE_nfe)$coef["X", "Pr(>|z|)"]
  
  # Anova-LRT
  anova_lrt_nfe <- anova(regXE_nfe, test="LRT")
  p_anova_lrt_nfe <- anova_lrt_nfe["X", "Pr(>Chi)"]
  
  # LRT
  chi_stat_lrt_nfe = 2*(logLik(regXE_nfe) - logLik(regE_nfe))
  p_lrt_nfe = 1-pchisq(chi_stat_lrt_nfe, df=1) # df??
  
  # Clean-up
  rm(regE_nfe, regXE_nfe, anova_lrt_nfe, chi_stat_lrt_nfe)
  
  # --- Record output to results table (p being converted to char) --- #
    result.mx <- rbind(result.mx, c(gene, 
      p_glm_kgen50, p_anova_lrt_kgen50, p_lrt_kgen50, p_glm_nfe, p_anova_lrt_nfe, p_lrt_nfe))
  
  # Clean-up
  rm(X, p_glm_kgen50, p_anova_lrt_kgen50, p_lrt_kgen50, p_glm_nfe, p_anova_lrt_nfe, p_lrt_nfe)
  
} # next variant

# Clean-up
rm(G, Y, E_kgen50, E_nfe, gene, sel_genes)

```

# explore_result

```{r explore_result}

result.df <- as.data.frame(result.mx)

result.df$gene <- as.vector(result.df$gene)

result.df$p_glm_kgen50 <- as.numeric(as.vector(result.df$p_glm_kgen50))
result.df$p_anova_lrt_kgen50 <- as.numeric(as.vector(result.df$p_anova_lrt_kgen50))
result.df$p_lrt_kgen50 <- as.numeric(as.vector(result.df$p_lrt_kgen50))

result.df$p_glm_nfe <- as.numeric(as.vector(result.df$p_glm_nfe))
result.df$p_anova_lrt_nfe <- as.numeric(as.vector(result.df$p_anova_lrt_nfe))
result.df$p_lrt_nfe <- as.numeric(as.vector(result.df$p_lrt_nfe))

str(result.df)

# Sort result
result.df <- result.df[order(result.df$p_lrt_nfe),]

# Look at correlation of diferent p-estimates
# Anova does not correlate with anything
# glm(wald?) more or less correlate with lrt

plot(p_glm_kgen50~p_anova_lrt_kgen50, data=result.df)
plot(p_glm_kgen50~p_lrt_kgen50, data=result.df)
plot(p_lrt_kgen50~p_anova_lrt_kgen50, data=result.df)

plot(p_glm_nfe~p_anova_lrt_nfe, data=result.df)
plot(p_glm_nfe~p_lrt_nfe, data=result.df)
plot(p_lrt_nfe~p_anova_lrt_nfe, data=result.df)

# Look at histograms of p-values
hist(result.df$p_glm_kgen50)
hist(result.df$p_anova_lrt_kgen50)
hist(result.df$p_lrt_kgen50)

sum(result.df$p_glm_kgen50 == 0)
sum(result.df$p_anova_lrt_kgen50 == 0)
sum(result.df$p_lrt_kgen50 == 0)

hist(result.df$p_glm_nfe)
hist(result.df$p_anova_lrt_nfe)
hist(result.df$p_lrt_nfe)

sum(result.df$p_glm_nfe == 0)
sum(result.df$p_anova_lrt_nfe == 0)
sum(result.df$p_lrt_nfe == 0)

# Load function for nice qq-plots
source("f03_qqplot.R")

qqunif.plot(result.df$p_glm_kgen50[!result.df$p_glm_kgen50 == 0], main="WECARE vs NFE (3 kgen50 eigenvectors, glm)")
qqunif.plot(result.df$p_anova_lrt_kgen50, main="WECARE vs NFE (3 kgen50 eigenvectors, anova_lrt)")
qqunif.plot(result.df$p_lrt_kgen50[!result.df$p_lrt_kgen50==0], main="WECARE vs NFE (3 kgen50 eigenvectors, lrt)")

qqunif.plot(result.df$p_glm_nfe[!result.df$p_glm_nfe == 0], main="WECARE vs NFE (3 nfe eigenvectors, glm)")
qqunif.plot(result.df$p_anova_lrt_nfe, main="WECARE vs NFE (3 nfe eigenvectors, anova_lrt)")
qqunif.plot(result.df$p_lrt_nfe, main="WECARE vs NFE (3 nfe eigenvectors, lrt)")

# Clean-up
rm(qqunif.plot, result.mx)

```

# data_summary

```{r data_summary}

dim(wecare_nfe_genotypes.mx)
class(wecare_nfe_genotypes.mx)
wecare_nfe_genotypes.mx[1:5,1:5]

dim(wecare_nfe_genes.df)
class(wecare_nfe_genes.df)
wecare_nfe_genes.df[1:5,]

dim(wecare_nfe_kgen.df)
colnames(wecare_nfe_kgen.df)
wecare_nfe_kgen.df[1:5,1:5]

dim(wecare_nfe_exac.df)
colnames(wecare_nfe_exac.df)
wecare_nfe_exac.df[1:5,1:5]

dim(wecare_nfe_variants.df)
str(wecare_nfe_variants.df)
wecare_nfe_variants.df[1:5,1:5]

dim(wecare_nfe_phenotypes.df)
str(wecare_nfe_phenotypes.df)
wecare_nfe_phenotypes.df[1:5,1:5]

dim(result.df)
str(result.df)
result.df[1:5,1:5]

# Check consistency of rownames and colnames
sum(colnames(wecare_nfe_genotypes.mx) != rownames(wecare_nfe_phenotypes.df))

sum(rownames(wecare_nfe_genotypes.mx) != rownames(wecare_nfe_kgen.df))
sum(rownames(wecare_nfe_genotypes.mx) != rownames(wecare_nfe_exac.df))
sum(rownames(wecare_nfe_genotypes.mx) != rownames(wecare_nfe_variants.df))

```

# save_data

```{r save_data}

save.image(paste(interim_data_folder, "s08a_wecare_vs_nfe_glm_3ev_Mar2017.RData", sep="/"))

```

# final_section

```{r final_section}

ls()
sessionInfo()
Sys.time()

```
