---
title: "nfe vs wecare: SKAT, 3ev nfe"
output: html_document
---

started: Alexey Larionov, 25Mar2017  
last updated: Alexey Larionov, 01Apr2017  

# Summary

Make QQ plot(s) for nfe vs wecare to see whether there is an inflation  
(recommended by DC during the Skype meeting of 14Mar2017)  

Use genes with > 2 variants in wecare-nfe  
23,523 -> 17,482 variants  
10,159 -> 5,324 genes  

Aggregate and calculate by SKAT-library  

Compare p-values generated by burden and SKAT algorithms  

Make QQ-plots using different functions  

# start_section

```{r start_section}

# Start time
Sys.time()

# Folders
#/analysis/mtgroup_share
#/scratch /medgen
setwd("/analysis/mtgroup_share/scripts/wecare_stat_03.17/scripts")
source_data_folder <- "/analysis/mtgroup_share/scripts/wecare_stat_03.17/source_data"
interim_data_folder <- "/analysis/mtgroup_share/scripts/wecare_stat_03.17/interim_data"
results_folder <- "/analysis/mtgroup_share/scripts/wecare_stat_03.17/results"

# Libraries & functions
library(SKAT)
source("f02_tryCatchAL.R") # for custom error handling
source("f03_qqplot.R") # for nice qq-plots

```

# load_data

```{r load_data}

# Load data
load(paste(interim_data_folder, "s07_filter_by_variant_effect_Mar2017.RData", sep="/"))

```

# check_data

```{r check_data}

dim(wecare_nfe_genotypes.mx)
class(wecare_nfe_genotypes.mx)
wecare_nfe_genotypes.mx[1:5,1:5]

dim(wecare_nfe_genes.df)
class(wecare_nfe_genes.df)
wecare_nfe_genes.df[1:5,]

dim(wecare_nfe_kgen.df)
colnames(wecare_nfe_kgen.df)
wecare_nfe_kgen.df[1:5,1:5]

dim(wecare_nfe_exac.df)
colnames(wecare_nfe_exac.df)
wecare_nfe_exac.df[1:5,1:5]

dim(wecare_nfe_variants.df)
str(wecare_nfe_variants.df)
wecare_nfe_variants.df[1:5,1:5]

dim(wecare_nfe_phenotypes.df)
str(wecare_nfe_phenotypes.df)
wecare_nfe_phenotypes.df[1:5,1:5]

# Check consistency of rownames and colnames
sum(colnames(wecare_nfe_genotypes.mx) != rownames(wecare_nfe_phenotypes.df))

sum(rownames(wecare_nfe_genotypes.mx) != rownames(wecare_nfe_kgen.df))
sum(rownames(wecare_nfe_genotypes.mx) != rownames(wecare_nfe_exac.df))
sum(rownames(wecare_nfe_genotypes.mx) != rownames(wecare_nfe_variants.df))

```

# select_variants

Should I use common - or at least present - in both?

Use genes with > 2 variants in wecare-nfe
23,523 -> 17,482 variants
10,159 -> 5,324 genes

```{r select_variants}

# --- Select variants for genes with > 2 aggregated events per dataset --- #

sel_genes <- wecare_nfe_genes.df[wecare_nfe_genes.df$counts_per_gene >= 3,"genes"]
length(sel_genes) # 5,324
sel_genes[1:5]

sel_vars <- wecare_nfe_variants.df[wecare_nfe_variants.df$SYMBOL %in% sel_genes, "SplitVarID"]
sel_vars <- as.character(as.vector(sel_vars))
length(sel_vars) # 17,842
sel_vars[1:5]

genotypes_sel.mx <- wecare_nfe_genotypes.mx[sel_vars,]
variants_sel.df <- wecare_nfe_variants.df[sel_vars,]

dim(genotypes_sel.mx)
dim(variants_sel.df)

# Clean-up
rm(sel_vars)

```

# prepare_data_for_analysis

Select the desired set of eigenvectors: nfe or nfe  

No missed data in eigenvectors  
Missed data in genotypes will be handled by SKAT  

```{r prepare_data_for_analysis}

# Eigenvectors
E <- as.matrix(wecare_nfe_phenotypes.df[,c("eig1_nfe","eig2_nfe","eig3_nfe")])
E[1:5,]
sum(is.na(E)) # 0

# Outcomes
Y <- ifelse(wecare_nfe_phenotypes.df$cc==-1, 0, 1)
sum(is.na(Y)) # 0
sum(Y) # 478
sum(!Y) # 198

# Genotypes: transpose
# Missed genotypes will be imputed by SKAT
G <- t(genotypes_sel.mx)
dim(G)
G[1:5,1:5]
sum(is.na(G)) # 887,225

# Clean-up
rm(genotypes_sel.mx)

```

# prepare_output_file

```{r prepare_output_file}

# Prepare header
header=c("gene", "num_vars", "svt_result", "p_svt", "MAP_svt", "burden_result", 
         "p_burden", "MAP_burden", "skat_result", "p_skat", "MAP_skat")

# Write header to output matrix 
# (do not use data frame here: it will play with rbind!)
result.mx <- matrix(ncol=length(header), nrow=0) 
colnames(result.mx) <- header

# Write header to output file
results_file="r08c_wecare_vs_nfe_skat_3ev_nfe_Mar2017.txt"
write(paste(header, sep="", collapse="\t"), 
      file=paste(results_folder, results_file, sep="/"))

# Clean-up
rm(header)

```

# SKAT

Skat warnings and errors are handled (except for the null model)

```{r SKAT}

# Calculate SKAT null model (the same for all genes)
skat_null <- SKAT_Null_Model(Y ~ E, out_type="D")
# D for the dichotomous outcome

# No warnings for kgen-50
# 2x warning messages for nfe:
# glm.fit: fitted probabilities numerically 0 or 1 occurred

#sel_genes <- sel_genes[2920:2930]

# For each gene
for(gene in sel_genes){
  
  #gene="ATM" # 16 variants
  #gene="CCNL2" # single variant 
  #gene <- sel_genes[850] # CRYGD - result=suceeded
  #gene <- sel_genes[2925] # SCUBE2 - double error (restarting interrupted promise evaluation)
  
  # Get variants
  vars <- as.vector(variants_sel.df[variants_sel.df$SYMBOL==gene, "SplitVarID"])
  
  # Get number of variants
  num_vars <- length(vars)

  # Get vector of genotypes  
  X <- G[,vars]
  
  # If there is one variant only
  if(num_vars == 1){

    # Aggregating variants (burden or skat) is irrelevant
    burden_result <- NA
    p_burden <- NA
    MAP_burden <- NA
    
    skat_result <- NA
    p_skat <- NA
    MAP_skat <- NA

    # --- Calculate test for single variant --- #
    svt <- tryCatchAL(SKATBinary_Single(X, skat_null))

    # Extract results of svt test
    # (see details in tryCarchAL function)
    
    if(svt$info == "error"){
      svt_result <- trimws(as.character(svt$error))
      p_svt <- NA
      MAP_svt <- NA
    }
    
    if(svt$info == "warning"){
      svt_result <- trimws(as.character(svt$warning))
      p_svt <- svt$value$p.value
      MAP_svt <- svt$value$MAP
    }

    if(svt$info == "succeeded"){
      svt_result <- "Succeeded"
      p_svt <- svt$value$p.value
      MAP_svt <- svt$value$MAP
    }
    
    # Clean-up
    rm(svt)
  }
  
  # If there are several variants in the gene  
  if(num_vars > 1){
    
    # Single-variant test is irrelevant
    svt_result <- NA
    p_svt <- NA
    MAP_svt <- NA
        
    # --- Calculate burden test --- #
    # default weights beta[1,25]
    burden <- tryCatchAL(SKATBinary(X, skat_null, method="Burden"))

    # Extract results of burden test
    # (see details in tryCarchAL function)
    
    if(burden$info == "error"){
      burden_result <- trimws(as.character(burden$error))
      p_burden <- NA
      MAP_burden <- NA
    }
    
    if(burden$info == "warning"){
      burden_result <- trimws(as.character(burden$warning))
      p_burden <- burden$value$p.value
      MAP_burden <- burden$value$MAP
    }

    if(burden$info == "succeeded"){
      burden_result <- "Succeeded"
      p_burden <- burden$value$p.value
      MAP_burden <- burden$value$MAP
    }
    
    # Clean-up
    rm(burden)
    
    # --- Calculate SKAT test --- #
    # default weights beta[1,25]
    skat <- tryCatchAL(SKATBinary(X, skat_null, method="SKAT"))

    # Extract results of skat test
    # (see details in tryCarchAL function)
    
    if(skat$info == "error"){
      skat_result <- trimws(as.character(skat$error))
      p_skat <- NA
      MAP_skat <- NA
    }
    
    if(skat$info == "warning"){
      skat_result <- trimws(as.character(skat$warning))
      p_skat <- skat$value$p.value
      MAP_skat <- skat$value$MAP
    }

    if(skat$info == "succeeded"){
      skat_result <- "Succeeded"
      p_skat <- skat$value$p.value
      MAP_skat <- skat$value$MAP
    }

    # Clean-up
    rm(skat)
  }

  # Add result to file
  write(paste(c(gene, num_vars, svt_result, p_svt, MAP_svt, burden_result, p_burden, 
                MAP_burden, skat_result, p_skat, MAP_skat), sep="", collapse="\t"), 
        file=paste(results_folder, results_file, sep="/"), append = TRUE)
  
  # Save results to matrix
  result.mx <- rbind(result.mx, c(gene, num_vars, svt_result, p_svt, MAP_svt, burden_result, 
                                  p_burden, MAP_burden, skat_result, p_skat, MAP_skat))
  
  # Clean-up
  rm(vars, num_vars, X, gene, svt_result, p_svt, MAP_svt, 
     burden_result, p_burden, MAP_burden, skat_result, p_skat, MAP_skat)
  
} # next gene

# Ceck result
dim(result.mx)

# Clean-up
rm(G, Y, E, sel_genes, variants_sel.df, skat_null, tryCatchAL, results_file)

```

# explore_result

```{r explore_result}

# Convert matrix to dataframe, 
# (change factors to char and numeric)
result.df <- as.data.frame(result.mx)

result.df$gene <- as.vector(result.df$gene)
result.df$num_vars <- as.numeric(as.vector(result.df$num_vars))
result.df$svt_result <- as.vector(result.df$svt_result)
result.df$p_svt <- as.numeric(as.vector(result.df$p_svt))
result.df$MAP_svt <- as.numeric(as.vector(result.df$MAP_svt))
result.df$burden_result <- as.vector(result.df$burden_result)
result.df$p_burden <- as.numeric(as.vector(result.df$p_burden))
result.df$MAP_burden <- as.numeric(as.vector(result.df$MAP_burden))
result.df$skat_result <- as.vector(result.df$skat_result)
result.df$p_skat <- as.numeric(as.vector(result.df$p_skat))
result.df$MAP_skat <- as.numeric(as.vector(result.df$MAP_skat))

str(result.df)

# Make combined scores
svt_burden_p <- apply(result.df[,c("p_svt", "p_burden")], 1, sum, na.rm=TRUE)
svt_burden_MAP <- apply(result.df[,c("MAP_svt", "MAP_burden")], 1, sum, na.rm=TRUE)

svt_skat_p <- apply(result.df[,c("p_svt", "p_skat")], 1, sum, na.rm=TRUE)
svt_skat_MAP <- apply(result.df[,c("MAP_svt", "MAP_skat")], 1, sum, na.rm=TRUE)

result.df <- cbind(result.df, svt_burden_p, svt_burden_MAP, svt_skat_p, svt_skat_MAP)
str(result.df)

# Sort result
result.df <- result.df[order(svt_burden_p),]

# Look at correlation of diferent p-estimates

plot(p_burden~p_skat, data=result.df)

# Look at histograms of p-values
hist(result.df$p_svt)
hist(result.df$p_burden)
hist(result.df$p_skat)
hist(svt_burden_p)
hist(svt_burden_MAP)

# Count NAs and extreme values
sum(is.na(result.df$p_svt))
sum(is.na(result.df$p_burden))
sum(is.na(result.df$p_skat))
sum(is.na(svt_burden_p))
sum(is.na(svt_burden_MAP))

sum(result.df$p_svt == 0, na.rm=TRUE)
sum(result.df$p_burden == 0, na.rm=TRUE)
sum(result.df$p_skat == 0, na.rm=TRUE)
sum(svt_burden_p == 0, na.rm=TRUE) # zeros for adding NAs
sum(svt_burden_MAP == 0, na.rm=TRUE) # zeros for adding NAs

sum(result.df$p_svt == 1, na.rm=TRUE)
sum(result.df$p_burden == 1, na.rm=TRUE)
sum(result.df$p_skat == 1, na.rm=TRUE)
sum(svt_burden_p == 1, na.rm=TRUE)
sum(svt_burden_MAP == 1, na.rm=TRUE)

# qqunif.plot function for nice qq-plots
source("f03_qqplot.R")

qqunif.plot(result.df$p_svt[!is.na(result.df$p_svt)], main="WECARE vs NFE (SKAT-svt, 3-nfe eigv)")
qqunif.plot(result.df$p_burden[!is.na(result.df$p_burden)], main="WECARE vs NFE (SKAT-burden, 3-nfe eigv)")
qqunif.plot(result.df$p_skat[!is.na(result.df$p_skat)], main="WECARE vs NFE (SKAT-skat, 3-nfe eigv)")

qqunif.plot(svt_burden_p[svt_burden_p != 0], main="WECARE vs NFE (SKAT svt-burden, 3-nfe eigv)")
qqunif.plot(svt_skat_p[svt_skat_p != 0], main="WECARE vs NFE (SKAT svt-skat, 3-nfe eigv)")

# SKAT qq-plot function with adjustment
QQPlot_Adj(result.df$p_svt, result.df$MAP_svt, Is.unadjsted=TRUE, 
           main="WECARE vs NFE (SKAT-svt, 3-nfe eigv, un-adj)")
QQPlot_Adj(result.df$p_burden, result.df$MAP_burden, Is.unadjsted=TRUE, 
           main="WECARE vs NFE (SKAT-burden, 3-nfe eigv, un-adj)")
QQPlot_Adj(result.df$p_skat, result.df$MAP_skat, Is.unadjsted=TRUE, 
           main="WECARE vs NFE (SKAT-skat, 3-nfe eigv, un-adj)")

QQPlot_Adj(svt_burden_p[svt_burden_p != 0], svt_burden_MAP[svt_burden_p != 0], Is.unadjsted=TRUE, 
           main="WECARE vs NFE (SKAT-svt-burden, 3-nfe eigv, un-adj)")
QQPlot_Adj(svt_skat_p[svt_skat_p != 0], svt_skat_MAP[svt_skat_p != 0], Is.unadjsted=TRUE, 
           main="WECARE vs NFE (SKAT-svt-skat, 3-nfe eigv, un-adj)")


QQPlot_Adj(result.df$p_svt, result.df$MAP_svt, Is.unadjsted=FALSE, 
           main="WECARE vs NFE (SKAT-svt, 3-nfe eigv, adj)")
QQPlot_Adj(result.df$p_burden, result.df$MAP_burden, Is.unadjsted=FALSE, 
           main="WECARE vs NFE (SKAT-burden, 3-nfe eigv, adj)")
QQPlot_Adj(result.df$p_skat, result.df$MAP_skat, Is.unadjsted=FALSE, 
           main="WECARE vs NFE (SKAT-skat, 3-nfe eigv, adj)")

QQPlot_Adj(svt_burden_p[svt_burden_p != 0], svt_burden_MAP[svt_burden_p != 0], Is.unadjsted=FALSE, 
           main="WECARE vs NFE (SKAT-svt-burden, 3-nfe eigv, adj)")
QQPlot_Adj(svt_skat_p[svt_skat_p != 0], svt_skat_MAP[svt_skat_p != 0], Is.unadjsted=FALSE, 
           main="WECARE vs NFE (SKAT-svt-skat, 3-nfe eigv, adj)")

# Clean-up
rm(qqunif.plot, result.mx, svt_burden_p, svt_burden_MAP, svt_skat_p, svt_skat_MAP)

```

# data_summary

```{r data_summary}

dim(wecare_nfe_genotypes.mx)
class(wecare_nfe_genotypes.mx)
wecare_nfe_genotypes.mx[1:5,1:5]

dim(wecare_nfe_genes.df)
class(wecare_nfe_genes.df)
wecare_nfe_genes.df[1:5,]

dim(wecare_nfe_kgen.df)
colnames(wecare_nfe_kgen.df)
wecare_nfe_kgen.df[1:5,1:5]

dim(wecare_nfe_exac.df)
colnames(wecare_nfe_exac.df)
wecare_nfe_exac.df[1:5,1:5]

dim(wecare_nfe_variants.df)
str(wecare_nfe_variants.df)
wecare_nfe_variants.df[1:5,1:5]

dim(wecare_nfe_phenotypes.df)
str(wecare_nfe_phenotypes.df)
wecare_nfe_phenotypes.df[1:5,1:5]

dim(result.df)
str(result.df)
result.df[1:5,1:5]

# Check consistency of rownames and colnames
sum(colnames(wecare_nfe_genotypes.mx) != rownames(wecare_nfe_phenotypes.df))

sum(rownames(wecare_nfe_genotypes.mx) != rownames(wecare_nfe_kgen.df))
sum(rownames(wecare_nfe_genotypes.mx) != rownames(wecare_nfe_exac.df))
sum(rownames(wecare_nfe_genotypes.mx) != rownames(wecare_nfe_variants.df))

```

# save_data

```{r save_data}

save.image(paste(interim_data_folder, "r08c_wecare_vs_nfe_skat_3ev_nfe_Mar2017.RData", sep="/"))

```

# final_section

```{r final_section}

ls()
sessionInfo()
Sys.time()

```
